### PR TITLE
Task2 Algorithm

### PR DESCRIPTION
**Pull request by JacobRajah**
Path optimization algorithm which will find the best route path to take based on reward, distance and time. The algorithm accounts for time to return to origin for battery swap (based on low battery calculation), time to takeoff/load/land, as well as distance between routes, including origin to the first waypoint. Algorithm will return a FlightPlan object containing all the flight information needed to complete the route and make estimates.

### CODE REVIEW

File: server/algorithm.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
+from waypoint import WAYPOINT_LST, Waypoint
+from route import Route
+from flightplan import FlightPlan
+
+
+def count_waypoint_occurances(curr_wp: Waypoint, final_waypoints: list[Waypoint]) -> int:
+    return len([wp for wp in final_waypoints if wp.name == curr_wp.name])
+
+
+def get_next_waypoint(current_waypoint: Waypoint, routes: list[Waypoint]) -> list[(Waypoint, int, bool)]:
```
Comment (CraigP17) [inline server/algorithm.py:17]: Docstring

File: server/algorithm.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    next_possible_waypoints = []
+    for i_route in range(len(routes)):
+        if routes[i_route].start_waypoint == current_waypoint:
+            next_possible_waypoints.append((routes[i_route].end_waypoint, i_route, True))
+        else:
+            next_possible_waypoints.append((routes[i_route].start_waypoint, i_route, False))
+    return next_possible_waypoints
+
+
+def compare_optimal_paths(path_1: FlightPlan, path_2: FlightPlan):
```
Comment (CraigP17) [inline server/algorithm.py]: Docstring

File: server/algorithm.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        return path_1
+    elif path_1.ratio == path_2.ratio:
+        # Equal path ratios, return path with smaller distance travelled
+        if path_1.distance_travelled < path_2.distance_travelled:
+            return path_1
+
+    return path_2
+
+
+def calculate_optimized_path(current_waypoint: Waypoint, routes: list[Waypoint], final_waypoints: list[Waypoint],
```
Comment (CraigP17) [inline server/algorithm.py]: Docstring

File: server/algorithm.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    flightplan = calculate_optimized_path(start_wp, all_routes, [start_wp], FlightPlan.time_to_takeoff)
+    flightplan.waypoints = [start_wp] + flightplan.waypoints
+    flightplan.takeoff()
+
+    print(flightplan.time_accumulated)
+    return flightplan.waypoints
+
+
+if __name__ == "__main__":
+    # sys.setrecursionlimit(5000)
```
Comment (CraigP17) [inline server/algorithm.py]: Comment needed?

File: server/flightplan.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
+from waypoint import WAYPOINT_LST, Waypoint
+from utils import calculate_distance
+
+
+class FlightPlan:
+    # Tunable Parameters
+    drone_speed = 10.0  # metres / seconds
```
Comment (CraigP17) [inline server/flightplan.py:7]: Note for future self, create issue for env file, and add these values to easily configure

**Comment by JacobRajah** [comment]
@CraigP17 Comments Resolved. Next steps include testing and modification for flight needs

**Review by CraigP17 — APPROVED** [review]
💯

**Pull request merged**